### DISCUSSION TITLE
How to override default gems with an older version (in particular Bundler)?

### DISCUSSION THREAD

**Discussion by sschuberth** [jruby/jruby#7403] [General]
I just realized JRuby 9.3.8.0 comes with [Bundler 2.2.29 as part of `default_gems`](https://github.com/jruby/jruby/blob/9.3.8.0/lib/pom.rb#L21). In my case, this seem to prevent working with Ruby projects that require an *older* version of Bundler. So, is there a way to override the default Bundler version? I'm currently using the `jruby-complete` artifact. Should I maybe [use `jruby-base` instead](https://github.com/jruby/jruby/issues/6205#issuecomment-693013027), and them somehow install Bundler manually?

PS: What's the difference between `default_gems` and `bundled_gems` in `pom.rb`? I'm asking because [this issue](https://github.com/jruby/jruby/issues/5686) is about not shipping Bundler as a bundled Gem, and I'm wondering why it's still a default Gem.

**Comment by enebo** [comment]
In C Ruby 2.6.x bundler is a default gem so we also do that (although we still support ruby 2.6.x while they don't so we have a newer version than the last version of C Ruby 2.6.x).    C Ruby 2.5.x did not install bundler as a default gem so that is the explanation for the issue you linked to.

jruby-complete is an all in one solution and you get what you get.  rubygems allows you to specify multiple directories for gems so you could install a newer bundler in a separate directory and it should use that version instead (assuming it is newer).  GEM_PATH??? I think is how you add an extra directory.  So set the right env and it should pick up gems from there too.

**Reply by sschuberth** [reply]
Yes, that's now it works for *newer* versions of Bundler: If I install a newer version of Bundler than the version that comes with jruby-complete, that newer version will be picked up. But the same does not work for older versions. How can I force to use my version of Bundler even if it's older then the one that comes with jruby-complete?

**Reply by enebo** [reply]
You may need to build jruby-complete with the version you need or use jruby-base and then add the gems you rely on.

**Reply by sschuberth** [reply]
Looks like I'm running into a kind of chicken-and-egg problem with jruby-base: As it does not seem to come with the `gem` tool, I cannot find a way to programmatically install bundler on top of it. I tried with (this is Kotlin code)

                org.jruby.Main().run(
                    arrayOf(
                        "-S", "gem", "install", "--no-document", "--user-install", "bundler:$bundlerVersion"
                    )
                )

but that fails with

```
LoadError: no such file to load -- rubygems
  require at org/jruby/RubyKernel.java:1017
   <main> at /usr/bin/gem:8
```

Any hint how to fix that / install a gem programmatically without the need to go through `org.jruby.Main().run` and use the `gem` tool?

**Reply by sschuberth** [reply]
> Any hint how to fix that / install a gem programmatically without the need to go through org.jruby.Main().run and use the gem tool?

I tried with

                val output = runScriptCode("""
                    require 'rubygems/commands/install_command'
                    cmd = Gem::Commands::InstallCommand.new
                    cmd.handle_options ["--no-document", "--user-install", "$BUNDLER_GEM_NAME:$bundlerVersion"]
                    cmd.execute
                """.trimIndent()).trim()

but that also does not work and fails with

```
(LoadError) no such file to load -- rubygems/commands/install_command
org.jruby.embed.EvalFailedException: (LoadError) no such file to load -- rubygems/commands/install_command
```

as jruby-base does not ship with rubygems.

**Reply by sschuberth** [reply]
So, can I maybe somehow modify the classpath / gempath for running scripts to contain my (older) bundler version before the default bundler version that ships with jruby-complete, so scripts would use my older bundler version instead of the default one?

**Reply by headius** [reply]
I am late to the conversation, but is there a reason your projects require an older Bundler? Could they not just be updated to a newer Bundler + lockfile?

The lack of RubyGems in jruby-base should be solvable by including the jruby-stdlib jar, though I'm not entirely sure including that jar would pull in the `gem` command. Could you give it a try?

Your use case is interesting... I would like to try to make it work!

**Reply by sschuberth** [reply]
> Could they not just be updated to a newer Bundler + lockfile?

Unfortunately not. I'm analyzing the dependencies of projects (by using the [OSS Review Toolkit](https://github.com/oss-review-toolkit/ort)) that are out of my control.

> The lack of RubyGems in jruby-base should be solvable by including the jruby-stdlib jar

I'll give that a try, thanks. However, isn't that then the same as using jruby-complete, as I thought that's exactly just jruby-base + jruby-stdlib, i.e. it would also include the newer Bundler version?

> Your use case is interesting... I would like to try to make it work!

Thank you!

**Reply by headius** [reply]
> isn't that then the same as using jruby-complete

jruby-complete = jruby-base + jruby-stdlib + other JRuby Java dependencies, so yes, I guess it would be similar. RubyGems is part of the standard library, so this would be the only way to use it from jruby-base.

**Reply by headius** [reply]
Oh, I think I see what you mean... if this is equivalent to jruby-complete, won't it bring along the newer Bundler.

Yes, it may initially. However if you use RubyGems to install a local copy of the older Bundler, I believe you can select that one from then on.

**Reply by sschuberth** [reply]
> However if you use RubyGems to install a local copy of the older Bundler, I believe you can select that one from then on.

How should that "selection" then work, programmatically? Just installing Bundler 1.0.0 does *not* work. Also with using jruby-base + jruby-stdlib (instead of jruby-complete), the newer Bundler version takes precedence over the older one:

```
INFO  org.ossreviewtoolkit.analyzer.managers.Bundler - Fetching bundler-1.0.0.gem
INFO  org.ossreviewtoolkit.analyzer.managers.Bundler - Successfully installed bundler-1.0.0
INFO  org.ossreviewtoolkit.analyzer.managers.Bundler - 1 gem installed
INFO  org.ossreviewtoolkit.analyzer.managers.Bundler - Installing the 'bundler' Gem in version 1.0.0 took 10.706047100s.
INFO  org.ossreviewtoolkit.analyzer.managers.Bundler - Using the 'bundler' Gem in version 2.2.29.
```

The last line is output from `puts(Gem::Specification.find_by_name('bundler').version)`.

**Reply by headius** [reply]
I think I figured out the magic sauce you need... bundler allows specifying the version to load at startup using `bundle _VERSION_ command`.

```
$ java -jar maven/jruby-complete/target/jruby-complete-9.3.9.0-SNAPSHOT.jar -S gem install bundler -v 1.0.0
2022-10-17T11:57:09.206-05:00 [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
Fetching bundler-1.0.0.gem
Successfully installed bundler-1.0.0
Parsing documentation for bundler-1.0.0
Installing ri documentation for bundler-1.0.0
Done installing documentation for bundler after 2 seconds
1 gem installed

$ java -jar maven/jruby-complete/target/jruby-complete-9.3.9.0-SNAPSHOT.jar -S bundle _1.0.0_ -v
2022-10-17T11:57:26.634-05:00 [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
Bundler version 1.0.0
```

You can also direct RubyGems to install and load from a specific path using `GEM_PATH`:

```
$ GEM_HOME=/tmp java -jar maven/jruby-complete/target/jruby-complete-9.3.9.0-SNAPSHOT.jar -S gem install bundler -v 1.0.0
2022-10-17T11:55:25.748-05:00 [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
Fetching bundler-1.0.0.gem
Successfully installed bundler-1.0.0
Parsing documentation for bundler-1.0.0
Installing ri documentation for bundler-1.0.0
Done installing documentation for bundler after 2 seconds
1 gem installed

[] ~/work/jruby93 $ GEM_HOME=/tmp java -jar maven/jruby-complete/target/jruby-complete-9.3.9.0-SNAPSHOT.jar -S bundle _1.0.0_ -v
2022-10-17T11:56:50.456-05:00 [main] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
Bundler version 1.0.0
```

**Reply by sschuberth** [reply]
> bundler allows specifying the version to load at startup using `bundle _VERSION_ command`.

Unfortunately, that does not help as I'm not calling the `bundle` CLI command directly. Instead, I'm analyzing Ruby projects that themselves depend on the Bundler Gem, and I need to ensure that these projects find Bundler 1.0.0 instead of Bundler 2.2.29.

**Comment by headius** [comment]
I have moved this to General discussions since there's more than just a simple Q&A here.
